processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -333,4 +348,73 @@ public boolean compareCannonicalValues(String s1, String s2)
       return s1 == s2;
     }
   }
+
+  // Caches references to selector objetcs for each column instead of creating a new object each time in order to save heap space.
+  private static class ObjectCachingColumnSelectorFactory implements ColumnSelectorFactory
+  {
+    private final ConcurrentMap<String, LongColumnSelector> longColumnSelectorMap = Maps.newConcurrentMap();
```

Comment:
@nishantmonu51 Currently `ColumnSelectorFactory.makeXXXXColumnSelector` – such as [this one](https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java#L507) – are not thread safe, this may be an oversight in the groupBy query engine. The resulting ColumnSelectors maybe be thread safe but not the methods that create them. We may want to investigate what needs to be done there.

Doesn't need to be for this PR, but I think the guarantees provided by the different methods are not clear. Maybe @cheddar has some input?